### PR TITLE
Add optional warnings when accessing properties and methods whose name starts with an underscore from other scripts.

### LINKED ISSUE #8095: Add optional warnings when accessing properties and methods whose name starts with an underscore from other scripts.
### Describe the project you are working on

Not related to a particular project, but could help people to keep a coherent style in GDScript. I was encouraged to open a proposal to discuss this idea when I opened godotengine/godot#83250 .

### Describe the problem or limitation you are having in your project

Whether GDScript should have access modifiers or not is a matter that has been discussed for a long time. Some people have suggested to add keywords or code annotations, but others are in favor of keeping everything public to avoid adding complexity to a language that was designed with simplicity in mind. Up to now, the convention has been to use identifier names starting with an underscore for private variables and functions.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Until the matter is resolved, I suggest to implement an optional warning (ignored by default) that triggers when an attribute or method whose name starts with an underscore is accessed from another script to enforce the convention, without adding complexity to the language.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

For example, this will trigger the warning, since `_private` and `_private_func` have a name that begins with an underscore and they are accessed from another script.
```
class_name MyScript
extends Node

var _private: int = 1

func _private_func():
    pass
```
```
extends Node

@onready var _my_script: MyScript = $MyScript

func _ready():
    _my_script._private += 1
    _my_script._private_func()
```
I should also mention that accessing the same variables and function from a child script will not trigger the warning, so technically they wouldn't be private, but protected. For example, this will not trigger the warning.
```
class_name OtherScript
extends MyScript

func _ready():
    _private += 1
    _private_func()
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This could probably be worked around by adding a `push_warning` statement to the setter function of some variables, but this could become cumbersome for scripts containing many private variables.

### Is there a reason why this should be core and not an add-on in the asset library?

Since this is also mentioned in the [official GDScript style guide](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html#functions-and-variables) I believe there should be an option in the engine by default.

### PR DESCRIPTION
**Pull request by HexagonNico**
Whether GDScript should have access modifiers or not is a matter that has been discussed for a long time. Up to now, the convention has been to use identifier names starting with an underscore for private variables and functions, so I suggest to implement an optional warning (ignored by default) that triggers when an attribute or method whose name starts with an underscore is accessed from another script.

For example, this will trigger the warning:
```gdscript
class_name MyScript
extends Node

var _private: int = 1

func _private_func():
    pass
```
```gdscript
extends Node

@onready var _my_script: MyScript = $MyScript

func _ready():
    _my_script._private += 1
    _my_script._private_func()
```

I believe this should help to enforce the convention.

I should also mention that accessing the same variables and function from a child script will not trigger the warning, so technically they wouldn't be private, but protected. For example, this will not trigger the warning.

```gdscript
class_name OtherScript
extends MyScript

func _ready():
    _private += 1
    _private_func()
```

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/8095*

### CODE REVIEW

**Comment by AThousandShips** [comment]
Please open a [proposal](https://github.com/godotengine/godot-proposals) to gauge the support for this idea

File: modules/gdscript/gdscript_analyzer.cpp [hunk 1]
```diff
@@ -3346,6 +3346,10 @@ void GDScriptAnalyzer::reduce_call(GDScriptParser::CallNode *p_call, bool p_is_a
 
 			parser->push_warning(p_call, GDScriptWarning::STATIC_CALLED_ON_INSTANCE, p_call->function_name, caller_type);
 		}
+
+		if (((String) p_call->function_name).begins_with("_")) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_analyzer.cpp:3350]:
```suggestion
		if (String(p_call->function_name).begins_with("_")) {
```

File: modules/gdscript/gdscript_analyzer.cpp [hunk 2]
```diff
@@ -4221,6 +4225,11 @@ void GDScriptAnalyzer::reduce_subscript(GDScriptParser::SubscriptNode *p_subscri
 				result_type = attr_type;
 				p_subscript->is_constant = p_subscript->attribute->is_constant;
 				p_subscript->reduced_value = p_subscript->attribute->reduced_value;
+#ifdef DEBUG_ENABLED
+				if (((String) p_subscript->attribute->name).begins_with("_")) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_analyzer.cpp:4229]:
```suggestion
				if (String(p_subscript->attribute->name).begins_with("_")) {
```

File: modules/gdscript/tests/scripts/analyzer/warnings/private_property_access.gd [hunk 3]
```diff
@@ -0,0 +1,10 @@
+func test():
+    var instance := TestClass.new()
+    instance.public += 1
+    instance._private += 1
+
+
+class TestClass:
+
+    var public := 1
+    var _private := 1
```
Comment (AThousandShips) [inline modules/gdscript/tests/scripts/analyzer/warnings/private_property_access.gd:10]:
```suggestion
    var _private := 1

```
Missing line termination

File: modules/gdscript/tests/scripts/analyzer/warnings/private_method_access.gd
Comment (AThousandShips) [inline modules/gdscript/tests/scripts/analyzer/warnings/private_method_access.gd:1]: These should use tabs for indentation

File: modules/gdscript/tests/scripts/analyzer/warnings/private_property_access.gd
Comment (AThousandShips) [inline modules/gdscript/tests/scripts/analyzer/warnings/private_property_access.gd:1]: Same here, tabs

**Comment by AThousandShips** [comment]
The documentation has to be in alphabetical order, update it by running `godot --doctool`, see [here](https://docs.godotengine.org/en/latest/contributing/documentation/updating_the_class_reference.html#updating-class-reference-when-working-on-the-engine)

**Comment by AThousandShips** [comment]
Also the test doesn't work correctly, it warns on access to ones in the same class or script, even without `self`

**Review by AThousandShips — CHANGES_REQUESTED** [review]
Needs to be rewritten to handle the access properly, currently it always throws the warning/error when a property/method with leading `_` is accessed, also needs more comprehensive tests to catch both false negatives and false positives, see above

**Comment by dalexeev** [comment]
> All of the effort to implement this really sounds like it could better be spent on actually bringing some form of a `private` keyword someday, huh?

I agree, if we don't want false positive warnings (see also https://github.com/godotengine/godot/issues/74879#issuecomment-1471634312), then untyped code should never produce the warning:

```gdscript
var a: OtherClass
var b: Variant
a._property # Warning.
b._property # No warning, can be the same class.
```

But then this warning can be ignored using `(expression as Variant)._property`, and will not cause an error at runtime, unlike real private variables.

This also needs to take inheritance into account, meaning that GDScript's "private" members are more like "protected" members (some languages, like Haxe, don't have a `protected` modifier either).

**Comment by HexagonNico** [comment]
Closing in favor of #641.

**Comment by AThousandShips** [comment]
Thank you for your contribution nonetheless!

**Pull request closed without merging**